### DISCUSSION TITLE
Scheduling the support-notifications service to send mail daily at 6am

### DISCUSSION THREAD

**Discussion by neerajasjawali** [edgexfoundry/community#107] [Q&A]
Hi All,

I am working on an use case where I have to send a mail notifications at scheduled time. Using EdgeX jakarta TLS.
For example, the scheduler service calls the core-data service to clean up the older events. Similarly, how do I make the scheduler service to call the support-notification service to send an email notification daily at 6 in the morning.? 

Thanks in advance.

**Comment by cloudxxx8** [comment]
This function is not supported right now.  It requires the Scheduler service further enhancement.
You can implement your own service to send the request by time, or just use the OS level scheduler, such as CRON, to send the Curl command to Notification service.

**Comment by neerajasjawali** [comment]
Hi @cloudxxx8,
Besides the default ScrubAged operation, are there any other operations where I can use the support-scheduler?

**Reply by cloudxxx8** [reply]
you can post more IntervalAction and Interval via its REST API
https://docs.edgexfoundry.org/2.3/api/support/Ch-APISupportScheduler/

**Comment by neerajasjawali** [comment]
@cloudxxx8, We can use support-notification only to send a message but not any attachment along with it. Am I right?

**Reply by cloudxxx8** [reply]
right

**Comment by neerajasjawali** [comment]
@cloudxxx8, can sensor value be sent via email notification using support-notification?

**Reply by cloudxxx8** [reply]
it's possible by using App Service to call the add notification API, but it would be better to implement Notifications service to receive data from message bus

**Reply by neerajasjawali** [reply]
> it's possible by using App Service to call the add notification API, but it would be better to implement Notifications service to receive data from message bus

@cloudxxx8 To do so, I need to add the MessageQueue section within the support-notifications config file same as defined in core-data config file. Correct?

**Reply by cloudxxx8** [reply]
no, it doesn't support this feature right now.  you need to modify the source code.

**Reply by neerajasjawali** [reply]
@cloudxxx8, Could you suggest me how to proceed with it?

**Reply by neerajasjawali** [reply]
> it's possible by using App Service to call the add notification API, but it would be better to implement Notifications service to receive data from message bus

@cloudxxx8, Could you explain in detail what you mean by "App Service to call the add notification API"?

**Reply by cloudxxx8** [reply]
Sensor values will be sent to message bus as Events/Readings, and App Service can receive them form the message bus.
Then, App Service has the capability to send REST call to create a new Notification in the Notification service.

**Reply by neerajasjawali** [reply]
> Sensor values will be sent to message bus as Events/Readings, and App Service can receive them form the message bus. Then, App Service has the capability to send REST call to create a new Notification in the Notification service.

@cloudxxx8 , App Service can only add the secrets and trigger using POST API. How can app service create a new notification? Is there any other API to call Notification API?

**Reply by cloudxxx8** [reply]
In your case, the trigger is message bus
https://docs.edgexfoundry.org/2.3/microservices/application/Triggers/#edgex-messagebus-trigger

You need to create a custom App Service with your own transform logic and convert the Event to Notification DTO, and then execute the Notifications POST API.  The custom App Service could be based on the SDK
https://docs.edgexfoundry.org/2.3/microservices/application/ApplicationFunctionsSDK/

**Reply by lenny-goodell** [reply]
Also you will want to use the NotificationsClient in your custom pipeline function.
https://docs.edgexfoundry.org/2.3/microservices/application/AppFunctionContextAPI/#notificationclient

Be sure to have the `Clients.support-notifications` in your configuration
https://docs.edgexfoundry.org/2.3/microservices/application/AppFunctionContextAPI/#note-about-clients

**Reply by neerajasjawali** [reply]
Thank you @cloudxxx8 , @lenny-intel. I will try implementing as suggested and get back to you. Thanks again.

**Reply by neerajasjawali** [reply]
> In your case, the trigger is message bus https://docs.edgexfoundry.org/2.3/microservices/application/Triggers/#edgex-messagebus-trigger
> 
> You need to create a custom App Service with your own transform logic and convert the Event to Notification DTO, and then execute the Notifications POST API. The custom App Service could be based on the SDK https://docs.edgexfoundry.org/2.3/microservices/application/ApplicationFunctionsSDK/

@cloudxxx8, how do I convert Event to Notification DTO?

**Reply by neerajasjawali** [reply]
> Also you will want to use the NotificationsClient in your custom pipeline function. https://docs.edgexfoundry.org/2.3/microservices/application/AppFunctionContextAPI/#notificationclient
> 
> Be sure to have the `Clients.support-notifications` in your configuration https://docs.edgexfoundry.org/2.3/microservices/application/AppFunctionContextAPI/#note-about-clients

@lenny-intel, how do I add the `NotificationClient() interfaces.NotificationClient` in the pipeline function?
I mean what are the parameters for notification client?

**Reply by neerajasjawali** [reply]
> > Also you will want to use the NotificationsClient in your custom pipeline function. https://docs.edgexfoundry.org/2.3/microservices/application/AppFunctionContextAPI/#notificationclient
> > Be sure to have the `Clients.support-notifications` in your configuration https://docs.edgexfoundry.org/2.3/microservices/application/AppFunctionContextAPI/#note-about-clients
> 
> @lenny-intel, how do I add the `NotificationClient() interfaces.NotificationClient` in the pipeline function? I mean what are the parameters for notification client?

@lenny-intel, Do I have to use the same parameters we use to define a notification post api for adding notification client in pipeline function?
For ex: 
```
[Writable.Pipeline.Functions.NotificationClient]
      [Writable.Pipeline.Functions.NotificationClient.Parameters]
      ID = ""
      Catergory = ""
      Content = ""
      ContentType = ""
      Description = ""
      Labels = ""
      Sender = ""
```

**Reply by lenny-goodell** [reply]
> > @lenny-intel, how do I add the `NotificationClient() interfaces.NotificationClient` in the pipeline function? I mean what are the parameters for notification client?

@neerajasjawali , you can not accomplish this with configurable pipeline in App Service Configurable. You must create you own custom app service and has a custom transform pipeline function that converts the Even/Ready to a Notification DTO and then sends it via the Notifications Client. 

See docs here for getting started with a custom application service.
docs.edgexfoundry.org/2.3/getting-started/ApplicationFunctionsSDK/

Also the quick start App Template is here:
https://github.com/edgexfoundry/app-functions-sdk-go/tree/v2.3/app-service-template#readme

**Reply by neerajasjawali** [reply]
> docs.edgexfoundry.org/2.3/getting-started/ApplicationFunctionsSDK/

@lenny-intel,
1)  As shown in the link (https://docs.edgexfoundry.org/2.1/getting-started/ApplicationFunctionsSDK/#:~:text=//%203)%20This%20is,1)%0A%20%20%20%20%7D), within the main.go file where we define the pipeline function configuration. I need to write a custom pipeline function for Notification. 
for ex: 
`transfrms.NewConversion().TransformToDTO`

2) To add the above pipeline function I have to define a function for TransformToDTO function within the given link
(https://github.com/edgexfoundry/app-functions-sdk-go/blob/jakarta/pkg/transforms/conversion.go)

3) And add the Notification Clients in app-service configuration.toml file.


Are these the correct approach to add the pipeline function?

**Reply by lenny-goodell** [reply]
Sort of. for step 2 you define your custom pipeline function in your own code.

Example for the App Template: https://github.com/edgexfoundry/app-functions-sdk-go/blob/main/app-service-template/functions/sample.go

**Reply by neerajasjawali** [reply]
> Sort of. for step 2 you define your custom pipeline function in your own code.
> 
> Example for the App Template: https://github.com/edgexfoundry/app-functions-sdk-go/blob/main/app-service-template/functions/sample.go

@lenny-intel, Do I need to write my own code in main.go file of app-service? If yes, cloud you also suggest me if there are any dependency files which I have to make modifications into? 
Or could you list out the files which I have to modify in order to create my custom app-service?

**Reply by lenny-goodell** [reply]
> Do I need to write my own code in main.go file of app-service?

Yes. The purpose to the App Template is for jump starting the creation of you custom app service.
Please read the documentation.
https://github.com/edgexfoundry/app-functions-sdk-go/tree/v2.3.1/app-service-template#readme

**Comment by neerajasjawali** [comment]
@cloudxxx8, I am trying to send a message which I have written it in the content section of notification. But in the mail the message is displayed as device-service is removed or created. How do I send the message of my choice using support-notification?

**Reply by cloudxxx8** [reply]
you might receive the mail from the core-metadata, not your notification content

**Comment by neerajasjawali** [comment]
Hi @cloudxxx8, Is it not possible to send a custom message via email using support-notification?

**Reply by cloudxxx8** [reply]
mail body should be the Notificaiont content, so it's custom text.

**Reply by neerajasjawali** [reply]
> mail body should be the Notificaiont content, so it's custom text.

@cloudxxx8, Now I am able to send the custom message, But along with the custom message, device-service status message is sent.
Do we receive an email notification only after the device-service is deleted or created? 

![Screenshot from 2023-05-12 15-21-50](https://github.com/edgexfoundry/community/assets/50900251/845112bc-a29e-4036-b923-fd943deb9a85)

**Reply by cloudxxx8** [reply]
no, it depends on your Subscription settings.

**Reply by neerajasjawali** [reply]
Okay. Thank you @cloudxxx8
